### ISSUE TITLE
JVM Json Serialization Broken with Array<Array<Primitive> types

### ISSUE THREAD

**Issue by AndrewReitz** [Kotlin/kotlinx.serialization#1446]
**Describe the bug**

Given the following code

```
@Serializable
    data class Matrix2(val data: Array<Array<Int>>) {
        override fun equals(other: Any?): Boolean {
            if (this === other) return true
            if (other == null || this::class != other::class) return false

            other as Matrix2

            if (!data.contentDeepEquals(other.data)) return false

            return true
        }

        override fun hashCode(): Int {
            return data.contentDeepHashCode()
        }
    }

    @Test
    fun test() {
        val m = Matrix2(
            arrayOf(
                arrayOf(1, 0, 0),
                arrayOf(0, 1, 0),
                arrayOf(0, 0, 1)
            )
        )
        val json = Json.encodeToString(m)
        val result: Matrix2 = Json.decodeFromString(json)

        assertEquals(m, result)
    }
```

The test should pass, but on the JVM it fails with the following error.

```
java.lang.ClassCastException: class [[Ljava.lang.Object; cannot be cast to class [[Ljava.lang.Integer; ([[Ljava.lang.Object; and [[Ljava.lang.Integer; are in module java.base of loader 'bootstrap')
	at serialization.JsonSerializationTest$Matrix2$$serializer.deserialize(JsonSerializationTest.kt:68)
	at serialization.JsonSerializationTest$Matrix2$$serializer.deserialize(JsonSerializationTest.kt:68)
	at kotlinx.serialization.json.internal.PolymorphicKt.decodeSerializableValuePolymorphic(Polymorphic.kt:63)
	at kotlinx.serialization.json.internal.StreamingJsonDecoder.decodeSerializableValue(StreamingJsonDecoder.kt:32)
	at kotlinx.serialization.json.Json.decodeFromString(Json.kt:85)
	at raytracer.serialization.JsonSerializationTest.test(JsonSerializationTest.kt:111)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
```

The generic type of Array<Array<Int> can be changed to any other primitive and the same behavior occurs. On js targets and native targets the code works as expected.

**Expected behavior**

Expected behavior would be that the Matrix2 class would be able to be serialized and deserialized on all target platforms.

**Environment**
kotlin = 1.4.32
kotlinx-serialization = 1.1.0
gradle = 7.0
jvm =  11.0.2 (Oracle Corporation 11.0.2+9)

**Comment by AndrewReitz** [comment]
A current work around is to switch the Arrays to Lists.

**Comment by AndrewReitz** [comment]
I didn't turn IR on. Is that the default for 1.4.32

**Comment by sandwwraith** [comment]
Nope. Then it's something new, thanks

**Comment by judos** [comment]
Workaround class I just wrote to keep my Array data model (serializer delegates to use List array instead):

Generic Class:
```
import kotlinx.serialization.KSerializer
import kotlinx.serialization.builtins.ListSerializer
import kotlinx.serialization.descriptors.SerialDescriptor
import kotlinx.serialization.encoding.Decoder
import kotlinx.serialization.encoding.Encoder

abstract class Array2DSerializer<S>(
		serializer: KSerializer<S>
) : KSerializer<Array<Array<S>>> {
	
	private val delegate = ListSerializer(ListSerializer(serializer))
	
	@Suppress("EXPERIMENTAL_API_USAGE")
	override val descriptor: SerialDescriptor = SerialDescriptor("Array2D", delegate.descriptor)
	
	override fun serialize(encoder: Encoder, value: Array<Array<S>>) {
		val list = value.map { it.toList() }
		delegate.serialize(encoder, list)
	}
	
	abstract fun createArr(w: Int, h: Int, list: List<List<S>>): Array<Array<S>>
	
	override fun deserialize(decoder: Decoder): Array<Array<S>> {
		val list = delegate.deserialize(decoder)
		return createArr(list.size, list[0].size, list)
	}
}
```

Serializer for specific type of Array<Array<BomberField>>:
```
import ch.judos.snakes.common.games.bomber.BomberField
import kotlinx.serialization.KSerializer
import kotlinx.serialization.serializer

/**
 * no clue why T and paramSerializer is used, but without it code does not compile
 */
class BomberFieldArray2DSerializer<T>(
		paramSerializer: KSerializer<T>
) : Array2DSerializer<BomberField>(serializer<BomberField>()) {
	
	override fun createArr(w: Int, h: Int, list: List<List<BomberField>>): Array<Array<BomberField>> {
		return Array(w) { x ->
			Array(h) { y ->
				list[x][y]
			}
		}
	}
}
```

Usage:
```
@kotlinx.serialization.Serializable
class BomberMap() {
	@kotlinx.serialization.Serializable(with = BomberFieldArray2DSerializer::class)
	lateinit var grid: Array<Array<BomberField>>
}
```

Some open questions I have:
1. is it possible to create Array of generic Type? `val x = arrayOf<S>() // Cannot use S as reified type param, use Class instead`
2. Why does any Serializer class require to have <T> and that serializer parameter? Without it I get java.lang.AssertionError: No such value argument slot in IrConstructorCallImpl: 0 (total=0).

Maybe it helps someone.

**Comment by qwwdfsad** [comment]
Fixed in Kotlin 1.5.0+
